File: prototype/js/yaml-parser.js
/** 将结构化 / 简写 YAML 统一为内部字段 */
export function normalizeYamlData(raw) {
  if (!raw || typeof raw !== 'object') return null;

  const isStructured = !!(raw.product || raw.prices || raw.promotions || raw.source || raw.sales);

  if (isStructured) {
    const product = raw.product || {};
    const prices = raw.prices || {};
    const source = raw.source || {};
    const sales = raw.sales || {};
    const promotions = Array.isArray(raw.promotions) ? raw.promotions : [];

    const title = product.title || product.name || '';
    const currentPrice = prices.discounted_price ?? prices.discountedPrice ?? prices.current_price ?? prices.currentPrice;
    const originalPrice = prices.original_price ?? prices.originalPrice;
    const currentDisplayPrice = prices.current_display_price ?? prices.currentDisplayPrice;
    const currency = prices.currency || 'CNY';
    const platform = source.platform || source.平台 || '京东';
    const logistics = source.logistics || source.物流 || '';
    const sold30Days = sales.sold_30_days ?? sales.sold30Days ?? sales.sold_30days ?? '';

    // 解析 promotions 自动拆分为标签和优惠券
    const { tags: autoTags, coupons: autoCoupons } = parsePromotions(promotions);

    // 合并根级 tags（AI prompt 输出的 tags[]）
    const rootTags = Array.isArray(raw.tags) ? raw.tags.filter(t => typeof t === 'string' && t.trim()) : [];
    const tags = [...new Set([...autoTags, ...rootTags])];

    // 合并根级 coupons（AI prompt 输出的 coupons[]）
    const rootCoupons = Array.isArray(raw.coupons) ? raw.coupons.map(c => ({
      count: Number(c.count) || 1,
      source: c.source || '',
      strength: c.strength || '',
      note: c.note || '',
    })).filter(c => c.source || c.strength) : [];
    const coupons = [...autoCoupons, ...rootCoupons];

    // 备注：优先使用根级 note，否则自动拼接
    const noteParts = [];
    if (raw.note && typeof raw.note === 'string' && raw.note !== 'null') noteParts.push(raw.note);
    if (currentDisplayPrice != null && Number(currentDisplayPrice) !== Number(currentPrice)) {
      noteParts.push(`展示价 ¥${currentDisplayPrice}`);
    }
    if (sold30Days) noteParts.push(`30天销量 ${sold30Days}`);

    const visual = parseVisual(raw.visual || product.visual || raw);
    return {
      title,
      currentPrice: numOrNull(currentPrice),
      originalPrice: numOrNull(originalPrice),
      currentDisplayPrice: numOrNull(currentDisplayPrice),
      currency: currency === 'CNY' ? '¥' : currency,
      platform,
      logistics,
      sales: sold30Days ? { sold30Days: String(sold30Days) } : null,
      category: product.category || product.分类 || inferCategory(title),
      link: source.link || source.url || product.link || '',
      note: noteParts.join(' · ') || '',
      tags,
      coupons,
      promotions: promotions.map(p => typeof p === 'string' ? p : (p.text || p.name || p.strength || '')).filter(Boolean),
      createdAt: raw.create_time || raw.createTime || null,
      ...visual,
    };
  }

  // 简写兼容格式
  const tags = Array.isArray(raw.tags) ? raw.tags : (raw.tags ? String(raw.tags).split(/[,，]/).map(t => t.trim()).filter(Boolean) : []);
  const coupons = Array.isArray(raw.coupons) ? raw.coupons.map(c => ({
    count: Number(c.count) || 1,
    source: c.source || '',
    strength: c.strength || c.discount || '',
    note: c.note || '',
  })) : [];
  if (Array.isArray(raw.promotions) && !raw.coupons) {
    const parsed = parsePromotions(raw.promotions);
    tags.push(...parsed.tags.filter(t => !tags.includes(t)));
    coupons.push(...parsed.coupons);
  }

  const promoList = Array.isArray(raw.promotions)
    ? raw.promotions.map(p => typeof p === 'string' ? p : (p.text || p.name || p.strength || '')).filter(Boolean)
    : [];
  const visual = parseVisual(raw.visual || raw);

  return {
    title: raw.title || raw.name || raw.product?.title || '',
    currentPrice: numOrNull(raw.currentPrice ?? raw.price ?? raw.现价 ?? raw.prices?.discounted_price),
    originalPrice: numOrNull(raw.originalPrice ?? raw.原价 ?? raw.prices?.original_price),
    currentDisplayPrice: numOrNull(raw.currentDisplayPrice ?? raw.prices?.current_display_price),
    currency: raw.currency || '¥',
    platform: raw.platform || raw.平台 || raw.source?.platform || '京东',
    logistics: raw.logistics || raw.source?.logistics || '',
    sales: raw.sales || null,
    category: raw.category || raw.分类 || inferCategory(raw.title || ''),
    link: raw.link || raw.url || raw.链接 || '',
    note: raw.note || raw.备注 || '',
    tags,
    coupons,
    promotions: promoList,
    createdAt: raw.create_time || raw.createTime || null,
    ...visual,
  };
}

/** 解析 visual / image_url / ascii_art */
export function parseVisual(raw) {
  const visual = raw.visual || {};
  const imageUrl = visual.image_url || visual.imageUrl || raw.image_url || raw.imageUrl || '';
  const asciiArt = visual.ascii_art || visual.asciiArt || raw.ascii_art || raw.asciiArt || '';
  let visualType = visual.type || raw.visualType || raw.visual_type || 'none';
  if (imageUrl) visualType = 'image';
  else if (asciiArt) visualType = 'ascii';
  return {
    visualType,
    imageUrl: imageUrl || '',
    asciiArt: asciiArt || '',
    image: '', // 图片 URL 需异步下载后填充
  };
}

export function numOrNull(v) {
  if (v == null || v === '') return null;
  const n = Number(v);
  return isNaN(n) ? null : n;
}

export function inferCategory(title) {
  if (/浴室|马桶|淋浴|家居|柜|床|沙发|收纳/.test(title)) return '家居';
  if (/手机|电脑|耳机|数码|iPhone|索尼|Apple/.test(title)) return '数码';
  if (/吸尘|洗衣|空调|冰箱|家电|戴森/.test(title)) return '家电';
  if (/面霜|精华|美妆|护肤|口红/.test(title)) return '美妆';
  if (/坚果|零食|食品|牛奶/.test(title)) return '食品';
  if (/衣|裤|鞋|服饰/.test(title)) return '服饰';
  return '其他';
}

/** 将 promotions 字符串拆分为标签与优惠券 */
export function parsePromotions(promotions) {
  const tags = [];
  const coupons = [];
  const serviceKw = /免费安装|质保|保价|包邮|赠运|售后/;
  const discountKw = /券|减|折|直降|到手价|满|赠送|PLUS|立减|秒杀/i;

  for (const item of promotions) {
    const text = typeof item === 'string' ? item.trim() : (item.strength || item.text || item.name || '').trim();
    if (!text) continue;

    if (serviceKw.test(text) && !/减|折|券/.test(text)) {
      tags.push(text);
    } else if (discountKw.test(text)) {
      let source = '促销';
      if (/券|PLUS/i.test(text)) source = '优惠券';
      else if (/直降|官方/i.test(text)) source = '官方直降';
      else if (/满|赠送/i.test(text)) source = '满赠活动';
      coupons.push({ count: 1, source, strength: text, note: '' });
    } else {
      tags.push(text);
    }
  }
  return { tags, coupons };
}

File: lib/core/utils/yaml_parser.dart
// YAML 优惠信息解析器
//
// 将 YAML 格式的优惠信息解析为结构化的 ParsedDeal 对象。
// 支持结构化格式（product / prices / source / promotions / coupons / visual）
// 和扁平格式（title / price / platform 等顶级字段）。
// 自动从促销权益中提取标签和优惠券信息。

import 'package:yaml/yaml.dart';
import 'package:uuid/uuid.dart';

/// 解析后的优惠信息
class ParsedDeal {
  /// 唯一标识（自动生成 UUID）
  final String id;
  /// 商品名称
  final String title;
  /// 来源平台（京东/淘宝/拼多多等）
  final String platform;
  /// 商品分类（数码/美妆/家电等）
  final String category;
  /// 到手价
  final double currentPrice;
  /// 原价
  final double? originalPrice;
  /// 页面展示价
  final double? displayPrice;
  /// 货币符号
  final String currency;
  /// 折扣描述（如 "6.6折"）
  final String? discount;
  /// 物流信息
  final String? logistics;
  /// 购买链接
  final String? link;
  /// 备注
  final String? note;
  /// 视觉内容类型（none/image/ascii）
  final String visualType;
  /// ASCII 艺术图内容
  final String? asciiArt;
  /// 图片 URL
  final String? imageUrl;
  /// 销量信息
  final String? sales;
  /// 创建时间（ISO8601 字符串，可选）
  final DateTime? createdAt;
  /// 来源信息 JSON（如 '{"sourceType":"手动新增","sourceRemark":null}'）
  final String? sourceJson;
  /// 标签列表
  final List<String> tags;
  /// 促销权益列表
  final List<String> promotions;
  /// 优惠券列表
  final List<ParsedCoupon> coupons;

  ParsedDeal({
    required this.id,
    required this.title,
    this.platform = '其他',
    this.category = '其他',
    required this.currentPrice,
    this.originalPrice,
    this.displayPrice,
    this.currency = '¥',
    this.discount,
    this.logistics,
    this.link,
    this.note,
    this.visualType = 'none',
    this.asciiArt,
    this.imageUrl,
    this.sales,
    this.createdAt,
    this.sourceJson,
    this.tags = const [],
    this.promotions = const [],
    this.coupons = const [],
  });
}

/// 解析后的优惠券信息
class ParsedCoupon {
  /// 券数量
  final int count;
  /// 券来源（店铺券/平台券/直播间等）
  final String source;
  /// 优惠力度（满300减50 / 9折 / 直减100等）
  final String strength;
  /// 备注（券码、领取方式等）
  final String? note;

  ParsedCoupon({
    this.count = 1,
    this.source = '',
    this.strength = '',
    this.note,
  });
}

/// YAML 优惠信息解析器
///
/// 将 YAML 字符串解析为 ParsedDeal 对象，支持结构化和扁平两种格式。
/// 自动从促销权益中提取标签和优惠券。
class YamlParser {
  static const _uuid = Uuid();

  /// Parse YAML string into a ParsedDeal
  static ParsedDeal parse(String yamlStr) {
    final dynamic doc = loadYaml(yamlStr);
    if (doc is! Map) {
      throw FormatException('YAML must be a map, got ${doc.runtimeType}');
    }

    final map = Map<String, dynamic>.from(doc);

    // Extract fields with both structured and flat format support
    final title = _extractTitle(map);
    final currentPrice = _extractCurrentPrice(map);
    final platform = _extractString(map, ['source.platform', 'platform', '平台']) ?? '其他';
    final category = _extractString(map, ['product.category', 'category', '分类']) ?? '其他';
    final originalPrice = _extractDouble(map, ['prices.original_price', 'originalPrice', '原价']);
    final displayPrice = _extractDouble(map, ['prices.current_display_price', 'currentDisplayPrice']);
    final currency = _extractCurrency(map);
    final logistics = _extractString(map, ['source.logistics', 'logistics']);
    final link = _extractString(map, ['source.link', 'link', 'url', '链接']);
    final note = _extractString(map, ['note', '备注']);
    final discount = _extractString(map, ['discount', '折扣']);
    final sales = _extractSales(map);
    final createdAt = _extractCreatedAt(map);

    // Visual
    final visualType = _extractVisualType(map);
    final imageUrl = _extractString(map, ['visual.image_url', 'image_url', 'imageUrl']);
    final asciiArt = _extractString(map, ['visual.ascii_art', 'ascii_art', 'asciiArt']);

    // Promotions, tags, coupons
    final promotions = _extractPromotions(map);
    final tags = _extractTags(map, promotions);
    final coupons = _extractCoupons(map, promotions);
    final sourceJson = _extractSourceJson(map);

    return ParsedDeal(
      id: _uuid.v4(),
      title: title,
      platform: platform,
      category: category,
      currentPrice: currentPrice,
      originalPrice: originalPrice,
      displayPrice: displayPrice,
      currency: currency,
      discount: discount,
      logistics: logistics,
      link: link,
      note: note,
      visualType: visualType,
      asciiArt: asciiArt,
      imageUrl: imageUrl,
      sales: sales,
      createdAt: createdAt,
      sourceJson: sourceJson,
      tags: tags,
      promotions: promotions,
      coupons: coupons,
    );
  }

  static String _extractTitle(Map<String, dynamic> map) {
    // Structured: product.title
    if (map['product'] is Map) {
      final p = Map<String, dynamic>.from(map['product']);
      if (p['title'] != null) return p['title'].toString();
    }
    // Flat
    for (final key in ['title', 'name', '商品名称']) {
      if (map[key] != null) return map[key].toString();
    }
    throw FormatException('Missing required field: title');
  }

  static double _extractCurrentPrice(Map<String, dynamic> map) {
    // Structured: prices.discounted_price
    if (map['prices'] is Map) {
      final p = Map<String, dynamic>.from(map['prices']);
      for (final key in ['discounted_price', 'current_price']) {
        if (p[key] != null) return _toDouble(p[key]);
      }
    }
    // Flat
    for (final key in ['currentPrice', 'price', '现价']) {
      if (map[key] != null) return _toDouble(map[key]);
    }
    throw FormatException('Missing required field: currentPrice');
  }

  static String? _extractString(Map<String, dynamic> map, List<String> keys) {
    for (final key in keys) {
      if (key.contains('.')) {
        final parts = key.split('.');
        dynamic current = map;
        bool found = true;
        for (final part in parts) {
          if (current is Map) {
            final m = Map<String, dynamic>.from(current);
            if (m.containsKey(part)) {
              current = m[part];
            } else {
              found = false;
              break;
            }
          } else {
            found = false;
            break;
          }
        }
        if (found && current != null) return current.toString();
      } else {
        if (map[key] != null) return map[key].toString();
      }
    }
    return null;
  }

  static double? _extractDouble(Map<String, dynamic> map, List<String> keys) {
    for (final key in keys) {
      if (key.contains('.')) {
        final parts = key.split('.');
        dynamic current = map;
        bool found = true;
        for (final part in parts) {
          if (current is Map) {
            final m = Map<String, dynamic>.from(current);
            if (m.containsKey(part)) {
              current = m[part];
            } else {
              found = false;
              break;
            }
          } else {
            found = false;
            break;
          }
        }
        if (found && current != null) return _toDouble(current);
      } else {
        if (map[key] != null) return _toDouble(map[key]);
      }
    }
    return null;
  }

  static String _extractCurrency(Map<String, dynamic> map) {
    String? raw;
    if (map['prices'] is Map) {
      final p = Map<String, dynamic>.from(map['prices']);
      raw = p['currency']?.toString();
    }
    raw ??= map['currency']?.toString();

    if (raw == null) return '¥';
    switch (raw.toUpperCase()) {
      case 'CNY':
      case 'RMB':
      case 'CNY/RMB':
        return '¥';
      case 'USD':
        return '\$';
      case 'EUR':
        return '€';
      case 'GBP':
        return '£';
      case 'JPY':
        return '¥';
      default:
        return raw;
    }
  }

  static String? _extractSales(Map<String, dynamic> map) {
    if (map['sales'] is Map) {
      final s = Map<String, dynamic>.from(map['sales'] as Map);
      return s['sold_30_days']?.toString() ?? s['sold30Days']?.toString();
    }
    return map['sales']?.toString();
  }

  static DateTime? _extractCreatedAt(Map<String, dynamic> map) {
    final raw = _extractString(map, ['created_at', 'createdAt']);
    if (raw == null || raw.isEmpty) return null;
    return DateTime.tryParse(raw);
  }

  static String _extractVisualType(Map<String, dynamic> map) {
    if (map['visual'] is Map) {
      final v = Map<String, dynamic>.from(map['visual'] as Map);
      if (v['type'] != null) return v['type'].toString();
    }
    if (map['visualType'] != null) return map['visualType'].toString();

    // Infer from content
    if (_extractString(map, ['visual.image_url', 'image_url', 'imageUrl']) != null) return 'image';
    if (_extractString(map, ['visual.ascii_art', 'ascii_art', 'asciiArt']) != null) return 'ascii';
    return 'none';
  }

  static List<String> _extractPromotions(Map<String, dynamic> map) {
    if (map['promotions'] is List) {
      return (map['promotions'] as List).map((e) => e.toString()).toList();
    }
    return [];
  }

  static List<String> _extractTags(Map<String, dynamic> map, List<String> promotions) {
    // Explicit tags
    if (map['tags'] is List) {
      return (map['tags'] as List).map((e) => e.toString()).toList();
    }

    // Auto-extract from promotions
    final tagKeywords = ['安装', '质保', '保价', '包邮', '免邮', '保修', '保障'];
    final tags = <String>[];
    for (final promo in promotions) {
      if (tagKeywords.any((kw) => promo.contains(kw))) {
        tags.add(promo);
      }
    }
    return tags;
  }

  static List<ParsedCoupon> _extractCoupons(Map<String, dynamic> map, List<String> promotions) {
    // Explicit coupons
    if (map['coupons'] is List) {
      return (map['coupons'] as List).map((e) {
        if (e is Map) {
          final m = Map<String, dynamic>.from(e);
          return ParsedCoupon(
            count: m['count'] != null ? (m['count'] as num).toInt() : 1,
            source: m['source']?.toString() ?? '',
            strength: m['strength']?.toString() ?? '',
            note: m['note']?.toString(),
          );
        }
        return ParsedCoupon(strength: e.toString());
      }).toList();
    }

    // Auto-extract from promotions
    final couponKeywords = ['券', '减', '折', '直降', '到手价', '满', '赠送', 'PLUS'];
    final coupons = <ParsedCoupon>[];
    for (final promo in promotions) {
      if (couponKeywords.any((kw) => promo.contains(kw))) {
        coupons.add(ParsedCoupon(
          source: '促销',
          strength: promo,
          count: 1,
        ));
      }
    }
    return coupons;
  }

  static String? _extractSourceJson(Map<String, dynamic> map) {
    String? sourceType;
    String? sourceRemark;

    // 结构化 source.type / source.remark
    if (map['source'] is Map) {
      final s = Map<String, dynamic>.from(map['source']);
      sourceType ??= s['type']?.toString();
      sourceRemark ??= s['remark']?.toString();
    }

    // 根级简写
    sourceType ??= _extractString(map, ['sourceType', '来源类型']);
    sourceRemark ??= _extractString(map, ['sourceRemark', '来源备注']);

    if (sourceType == null && sourceRemark == null) return null;

    final buffer = StringBuffer();
    buffer.write('{');
    if (sourceType != null) {
      buffer.write('"sourceType":"$sourceType"');
      if (sourceRemark != null) {
        buffer.write(',"sourceRemark":"$sourceRemark"');
      }
    } else if (sourceRemark != null) {
      buffer.write('"sourceRemark":"$sourceRemark"');
    }
    buffer.write('}');
    return buffer.toString();
  }

  static double _toDouble(dynamic value) {
    if (value is num) return value.toDouble();
    if (value is String) {
      final cleaned = value.replaceAll(RegExp(r'[^\d.\-]'), '');
      return double.tryParse(cleaned) ?? 0.0;
    }
    return 0.0;
  }
}
